File: m2-h2/src/main/java/ru/yegorpilipenko/otus/spring/m2h2/service/BookService.java
package ru.yegorpilipenko.otus.spring.m2h2.service;

import ru.yegorpilipenko.otus.spring.m2h2.document.Author;
import ru.yegorpilipenko.otus.spring.m2h2.document.Book;
import ru.yegorpilipenko.otus.spring.m2h2.document.Genre;

import javax.annotation.Nonnull;
import java.util.Collection;
import java.util.Optional;

public interface BookService {

    @Nonnull
    Collection<Book> getAll();

    @Nonnull
    Optional<Book> getById(@Nonnull String id);

    @Nonnull
    Book save(@Nonnull String name, @Nonnull Collection<Author> authors, @Nonnull Collection<Genre> genres);

    boolean remove(@Nonnull String id);

}

File: m2-h1/src/main/kotlin/ru/yegorpilipenko/otus/spring/m2h1/service/BookService.kt
package ru.yegorpilipenko.otus.spring.m2h1.service

import ru.yegorpilipenko.otus.spring.m2h1.entity.Book

interface BookService {

    fun findAll(): Collection<Book>

    fun findById(id: Long): Book?

    fun save(name: String, authorIds: Collection<Long>, genreIds: Collection<Long>): Book

    fun removeById(id: Long): Boolean

}
